File: Cuarto/Programación Distribuida/java_concurrencia3/ejercicio4.java
import java.util.concurrent.*;

public class ejercicio4 { 
    static BlockingQueue<Integer> q;

    public static int fib(int n) {
        if (n == 0 || n == 1) return n;
        else return fib(n-1) + fib(n-2);
    }

    public static class Productor extends Thread {
        int n;
        public Productor(int n) {
            this.n = n;
        }

        public void run(){
            for (int i = 0; i < n; i++) {
                try {
                    sleep(500);
                    q.put(fib(i));
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class Consumidor extends Thread {
        public Consumidor() {}

        public void run(){
            while(true){
                try {
                    sleep(1000);
                    Integer fib = q.poll();
                    if (fib == null) {
                        System.out.println("No hay elementos en la cola");
                    } else {
                        System.out.println(fib);
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static void main(String[] args) {
        q = new SynchronousQueue<>();
        Thread t = new Productor(10);
        t.start();
        Thread t2 = new Consumidor();
        t2.start();
    }
}

File: Cuarto/Programación Distribuida/python_concurrencia3/ejercicio4.py
import threading
import time
import queue

def fib(n):
    if n == 1 or n == 0:
        return n
    else:
        return fib(n-1) + fib(n-2)

class Productor(threading.Thread):
    def __init__(self, n, cola):
        super().__init__()
        self.cola = cola
        self.n = n
    
    def run(self):
        for i in range(0, self.n):
            time.sleep(2)
            self.cola.put(fib(i))

class Consumidor(threading.Thread):
    def __init__(self, cola):
        super().__init__()
        self.cola = cola
    
    def run(self):
        while True:
            time.sleep(1)
            try:
                print(self.cola.get_nowait())
            except:
                print("No hay elementos en la cola")

cola = queue.Queue()
productor = Productor(10, cola)
consumidor = Consumidor(cola)

productor.start()
consumidor.start()